File: 프로그래머스/0/120880. 특이한 정렬/특이한 정렬.cpp
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

bool compare(pair<int,int>a,pair<int,int>b)
{
        if (a.first == b.first) 
        {
            return a.second > b.second;
        }
        return a.first < b.first;
}

vector<int> solution(vector<int> numlist, int n) {
    vector<int> answer;

    vector<pair<int, int>> tmp;

    for (int i = 0; i < numlist.size(); ++i) {
        int distance = abs(numlist[i] - n);
        tmp.push_back({distance, numlist[i]});
    }

    sort(tmp.begin(), tmp.end(), compare);

    for (auto result : tmp) {
        answer.push_back(result.second);
    }

    return answer;
}

File: 프로그래머스/0/120880. 특이한 정렬/특이한 정렬.py
def solution(numlist, n):
    # 정렬 기준:
    # 1. n과의 거리(절대값)를 기준으로 오름차순 정렬
    # 2. 거리가 같을 경우 숫자 자체를 내림차순으로 정렬
    return sorted(numlist, key=lambda x: (abs(x - n), -x))
